server/src/lsp-server.spec.ts

```diff
@@ -155,6 +155,41 @@ describe('diagnostics', () => {
         assert.equal("Cannot find name 'missing'.", fileDiagnostics[0].message);
     }).timeout(10000);
 
+    it('supports diagnostic tags', async () => {
+        const doc = {
+            uri: uri('diagnosticsBar.ts'),
+            languageId: 'typescript',
+            version: 1,
+            text: `
+        import { join } from 'path';
+        /**
+         * @deprecated
+         */
+        function foo(): void {
+        }
+
+        foo();
+      `
+        };
+        server.didOpenTextDocument({
+            textDocument: doc
+        });
+
+        await server.requestDiagnostics();
+        await new Promise(resolve => setTimeout(resolve, 200));
+        // assert.deepEqual(JSON.stringify(diagnostics, null, 2), '');
+        const resultsForFile = diagnostics.find(d => d!.uri === doc.uri);
+        assert.isDefined(resultsForFile);
+        const fileDiagnostics = resultsForFile!.diagnostics;
+        assert.equal(fileDiagnostics.length, 2);
+        const unusedDiagnostic = fileDiagnostics.find(d => d.code === 6133);
+        assert.isDefined(unusedDiagnostic);
+        assert.deepEqual(unusedDiagnostic!.tags, [lsp.DiagnosticTag.Unnecessary]);
```

Comment:
If you wanted to avoid the null assertions, this should be safe (but type checking of the params might fail?) - this will definitely fail when run if `undefined`.

```suggestion
        assert.deepEqual(unusedDiagnostic?.tags, [lsp.DiagnosticTag.Unnecessary]);
```

Reply:
This should be safe because the `assert.isDefined(unusedDiagnostic);` will fail the test before that line when `unusedDiagnostic` is undefined.

Reply:
Yeah, I just don't like leaving `!`s

Reply:
I think in this case it's very appropriate because we are expressing the fact that we know that it's defined at this point.